druid-shell/src/alert.rs

```diff
@@ -0,0 +1,451 @@
+    Error,
+}
+
+/// A specific button for the alert dialog.
+///
+/// If the alert is just reporting information and it only needs a single button,  
+/// then you should use a primary button with the label **OK** or a translation of it.
+///
+/// The cancel button should always be labeled **Cancel** or a translation of it.
+///
+/// In all other cases you should give a short but descriptive label that includes a verb.
+/// The label should let the user know what will happen when they click the button.
+/// Assume that the user didn't actually read the alert message. This assumption ends up
+/// true more often than not, because alerts by design appear unexpectedly while the user
+/// is trying to accomplish some other task.
+///
+/// Labels like **Send message**, **View all**, **Reconnect**, **Remind me tomorrow**, and
+/// **Delete file** make it much more clear what will happen compared to a generic **OK**,
+/// **Confirm**, or **Yes**.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AlertButton {
+    pub(crate) label: Cow<'static, str>,
+}
+
+impl AlertToken {
+    /// A token that does not correspond to any alert.
+    pub const INVALID: AlertToken = AlertToken(0);
+
+    /// Create a new token with the specified `id`.
+    pub(crate) const fn new(id: usize) -> AlertToken {
+        AlertToken(id)
+    }
+}
+
+impl AlertResponse {
+    /// Create a new alert response.
+    pub(crate) fn new(token: AlertToken, button: Option<AlertButton>) -> AlertResponse {
+        AlertResponse { token, button }
+    }
+
+    /// Returns the [`AlertToken`] that identifies the alert.
+    ///
+    /// [`AlertToken`]: struct.AlertToken.html
+    #[inline]
+    pub fn token(&self) -> AlertToken {
+        self.token
+    }
+
+    /// Returns the [`AlertButton`] that closed the alert.
+    ///
+    /// # Canceled
+    ///
+    /// This will return `None` if the alert was canceled.
+    ///
+    /// Keep in mind that the platform may forcefully cancel the alert, e.g. when
+    /// the parent window is closed, even when the alert dialog has no cancel button.
+    ///
+    /// [`AlertButton`]: struct.AlertButton.html
+    #[inline]
+    pub fn button(&self) -> Option<&AlertButton> {
+        self.button.as_ref()
+    }
+}
+
+impl AlertButton {
+    /// A primary button with the English label **OK**.
+    pub const OK: AlertButton = AlertButton::new("OK");
+    /// A cancel button with the English label **Cancel**.
+    pub const CANCEL: AlertButton = AlertButton::new("Cancel");
+
+    /// Create a new `const` button with the specified `label`.
+    ///
+    /// You should give a short but descriptive label like **Send message** that lets
+    /// the user know what will happen when they click the button.
+    ///
+    /// See [`AlertButton`] for more information.
+    ///
+    /// [`AlertButton`]: struct.AlertButton.html
+    pub const fn new(label: &'static str) -> AlertButton {
```

Comment:
Thinking out loud: at least from `druid` we're going to want to be (eventually) pushing localized text wherever possible, and I wonder if we should have like `static` and `new` instead of `new` and `dynamic`? That is, having the method folks are most likely to use be the flexible one.

I'm also not especially concerned about the allocation cost, given the use case.

Reply:
I am definitely open to this, in an earlier iteration I had `const_new` and `new`. My reasoning for changing it was that most apps aren't going to localize anything.

I think `static` is a reserved keyword just like `const`. So the method name would have to be something different. Any ideas?

Reply:
I think `const_new` is good enough.